#18 cleaning

matkrak opened the pull request:
Removed hardcoded metadata.
I see **str** are quite messy right now, I think a separate issue would be nice to handle this. Or maybe **str** is something that should be written at the very end when more methods will be available?

matkrak committed:
Removed hardcoded metadata

That was not a problem at all, I just wonder if __array_finalize__ should look like it does or it needs to be changed.

grzanka commented:
Speaking of str - I agree - it requires cleaning. Create separate issue for that.

grzanka commented:
What about **array_finalize** method ? 
When is it called ? Is it needed now ?
You can find some useful information here: http://docs.scipy.org/doc/numpy-1.10.1/user/basics.subclassing.html

I think few use cases of different methods of object creation in main method would help to clarify the picture - explicit constructor, view casting and new from template

matkrak commented:
Well I have already read this article you mentioned. I see <b>array_finalize</b> is used to copy extra information like metadata to an object that subclasses np.array.
Also I have already done it in <b>new</b> method and it works. 
Should I remove metadata initialization from <b>new</b> and move it do <b>array_finalize</b> instead? 
Could be useful if one wanted to create an object that sublacces curve somehow not by using explicit constructor but view casting or a template.

grzanka commented:
Before removing array_finalize, add some printouts to `__new__` and `__array_finalize__`.
Then add few examples of creating objects with different methods, described in the article I've mentioned. Use explicit constructor, slicing and view. Check if it works with all comination of subclasses, i.e.:

```
c1 = Curve(...)
c1.view(np.ndarray)

c2 = Curve(...)
c2.view(Profile)

c2 = Profile(...)
c2[:1,2:]
```

matkrak commented:
OK <b>array_finalize</b> is necessary to provide good behavior of objects. I will rewrite this method.

matkrak committed:
This time maybe it works

added just 1 line, but very important one.
Also in line 112 some changes just to see if metadata is bein copied

matkrak committed:
__array_finalize__ method changed

added just 1 line, but very important one.
Also in line 112 some changes just to see if metadata is bein copied

matkrak commented:
Now I see why it was so important. I hope I did it well, just one question.
Since no extra data is added do a Curve to make it a profile, <b>**array_finalize**</b> of Profile and LateralProfile doesn't need this extra line (it is being copied on Curve level). Still if there is any other data added 'extra' to a Profile or LateralProfile - it will affect <b>**array_finalize**</b> method. Am I right?

If you say it is done correctly I will remove unecessary comments I made for testing purpose.

grzanka commented:
Still some problems:

```
>>> from beprof import profile
>>> from beprof import curve
>>> c1 = curve.Curve( [[0,0],[1,1]], generator="sin")
>>> print(c1)
shape: (2, 2)
X : [0.000,1.000]
Y : [0.000000,1.000000]
Metadata : {'generator': 'sin'}
>>> 
>>> p1 = c1.view(profile.Profile)
>>> print(p1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/grzanka/tmp/beprof/beprof/profile.py", line 123, in __str__
    ret = super().__str__()
  File "/home/grzanka/tmp/beprof/beprof/curve.py", line 154, in __str__
    "\nMetadata : " + str(self.metadata)
AttributeError: 'Profile' object has no attribute 'metadata'

```

grzanka commented:
Trying to answer your question - the best would be to draw a diagram how `__new__` methods are called and try to do the printouts as you did in `__array_finalize__`

matkrak committed:
Cosmetical changes for testing purpose

Nothing importand has changed in this file

matkrak committed:
Trying console again

Nothing importand has changed in this file

matkrak committed:
Changed all array_finalize

Now it should all work as it is supposed to

matkrak committed:
Changed all array_finalize

Now it should all work as it is supposed to

matkrak commented:
I don't know why but I didn't consider one could view curve as a profile. So now it is possible, and required the same line in every <b>**array_finalize**</b> method.
Again, if it works OK I will remove unecessary 'prints' so that you can close and merge.

grzanka commented on beprof/curve.py:
```diff
@@ -51,8 +53,10 @@ def __new__(cls, input_array, **meta):
+        # print("Here I am in Curve.__array_finalize__ obj: ", type(obj))
+        if obj is None: # what generally means the object was created using explicit constructor
             return
+        self.metadata = getattr(obj, 'metadata', None)
```
There is an inconsistency when you are dealing with lack of metadata:

```
>>> import numpy as np
>>> from beprof import profile
>>> from beprof import curve
>>> n = np.asarray( [[0,0], [2,2]])
>>> nc = n.view(curve.Curve)
>>> print(n)
[[0 0]
 [2 2]]
>>> print(nc)
shape: (2, 2)
X : [0.000,2.000]
Y : [0.000000,2.000000]
Metadata : None
>>> nc.metadata
>>> c = curve.Curve( [[0,0], [1,1]])
>>> print(c)
shape: (2, 2)
X : [0.000,1.000]
Y : [0.000000,1.000000]
Metadata : {}
>>> c.metadata
{}
```

Shouldn't you rather take empty directory as default value, rather than none ?

Check also `Profile`

matkrak replied:
I agree, will change it in a minute

matkrak committed:
Changed default metadata to empty dict

matkrak committed:
Changed default metadata to empty dict

matkrak committed:
As in profile, changes default meta to {}

matkrak committed:
As in profile, changes default meta to {}

matkrak commented:
Anything to do here?

grzanka commented:
It is OK, I will merge it later

grzanka commented:
Merged

grzanka commented:
I am not sure if it was merged correctly, can you update your branch with new master ?

matkrak commented:
I have some problems merging it. Can we leave it for tomorrow, when I can pose questions via Skype? 
What is more all my branches need update now since grzanka-master was updated, and I am not really sure if I am doing it the correct way.

grzanka commented:
OK, lets postpone it. I also recommend to use PyCharm to solve conflicts in files (try to google it).

matkrak committed:
vol1 elo

matkrak committed:
can be repaired even if i break something

matkrak committed:
solving conflict i hope?

matkrak committed:
Might be necessary to add since I am using Pycharm now

matkrak committed:
Might be necessary to add since I am using Pycharm now

matkrak committed:
solving conflicts

matkrak committed:
Merge branch 'grzanka-master' into HEAD

matkrak committed:
Merge branch '#18-cleaning' of https://github.com/matkrak/beprof into #18-cleaning

matkrak committed:
continue cleaning

matkrak commented:
Ready to merge

grzanka commented:
Why the files are duplicated ? I see curve.py in main directory and also in beprof directory. 
Complete mess now.

matkrak committed:
rm 1

matkrak committed:
rm 2

matkrak committed:
rm 3

matkrak committed:
rm 4

matkrak committed:
rm 5

matkrak committed:
rm 6

matkrak committed:
rm 7

grzanka commented on beprof/curve.py:
```diff
@@ -54,6 +54,11 @@ def __array_finalize__(self, obj):
+        # print("Here I am in Curve.__array_finalize__ obj: ", type(obj))
+        if obj is None: # what generally means the object was created using explicit constructor
+            return
+        self.metadata = getattr(obj, 'metadata', {})
```
Why this is self.metadata is executed twice ?

grzanka commented:
The goal of this PR was to remove hardocoded metadata. When I look on "Files changed" tab I don't see this has happened. Maybe your "rm" commits removed it ?

matkrak committed:
Corrected doubled lines

I have no idea what happened

matkrak commented:
This is suprising, but hardcoded metadata has been removed anyway. Maybe It was removed "by mistake" in other pull request you already merged? A terrible mess with this repo, but now I know how to work with it.
